### ISSUE TITLE
lttree_postprocess.py error:  reference to undefined $angle

### ISSUE THREAD

**Issue by davidfir3** [jewettaij/moltemplate#48]
Hi Andrew.

I was trying to modified a .lt file generated by ATB to fit OPLSAA force field. However, an error occurred as below:
```
lttree_postprocess.py v0.6.1 2020-10-15

Error(lttree_postprocess.py): near "D7JH_allatom_optimized_geometry.lt", line 272

      Reference to undefined $angle:

            $/angle:bcb/a2     (<--full name)

      (If that angle belongs to a molecule (or other subunit), make sure that
       you specified the correct path which leads to it (using / and ..))
```

I don't think this error result from mistakes in the .lt file because the assignment for that angle can be found in "ttree_assignments.txt" file.
`$/angle:bcb/a2   2       #"D7JH_allatom_optimized_geometry.lt", line 272`

This error was omitted when running the script with "-nocheck" flag. However, there will be extra sections of "Angle List", "Dihedral List" and "Improper List" in the .data file.

Could you tell me what's the problem?
Thanks

**Comment by jewettaij** [comment]
Hi davidfir3
It's tricky to diagnose the problem without seeing your LT files.
If you send them to me by [email](http://moltemplate.org/images/author_email.png), I will take a look in the next day or two.
Andrew
P.S.  There may have been some changes to the format of the LT files generated by ATB in the last few days.  I asked them to change their file format recently.  However I don't think that this would be the reason for the problem you are having.

**Comment by jewettaij** [comment]
Thanks for sending me your files.
There are several problems:

The main problem is that moltemplate.sh is generating a misleading and confusing error message.  (You're right.  The "$/angle:bcb/a1" angle is defined.)  
There are problems with your files, but moltemplate.sh should do a better job explaining what those problems are.  The error message is extremely confusing.  
Thank you for reporting this.

Meanwhile, to get around this problem, remove the "Data Angle List", "Data Dihedral List", and "Data Improper List" sections from your "D7JH_allatom_optimized_geometry.lt" file.  Then try again.  

Note that, the files from the ATB web site include a list of angles, dihedrals, and impropers.
But if you are using the OPLSAA force field, then there is no need to specify them because (like most force fields) the "oplsaa.lt" file contains rules for generating them automatically.
*(Unfortunately those rules only work if you have selected your @atom types correctly.  (For OPLSAA, this can be difficult for a complicated molecule like yours.  (Discussion [here](http://moltemplate.org/force_field_recommendations.html).))*

May I ask, why are you trying to use OPLSAA with files generated by ATB?  The authors of the ATB service claim that their molecules have been optimized and are more accurate than molecules using any force field (including OPLSAA, AMBER/GAFF, GROMOS, CHARMM, etc...)  *(If I understand ATB correctly, then every atomic charge, bond, angle, dihedral, and improper in your molecule was selected to minimize the difference with explicit DFT results.)*

If, for some reason, you do want to manually specify the list of angles, dihedrals, and impropers, then you must change your files.  It's confusing because the "Data Bond List" section allows you to specify the atoms in each bond and omit the @bond type.  But there is nothing similar for angles, dihedrals, and impropers.  Instead you would create a "Data Angles" section with 5 columns, *including the @angle type:*
```
write("Data Angles") {
   $angle:a1   @angle:TYPE1   $atom:H30    $atom:C18    $atom:H28    
   $angle:a2   @angle:TYPE2   $atom:H30    $atom:C18    $atom:H29  
    :
}
```
...(where "TYPE1" and "TYPE2" refer to angle types defined in the "oplsaa.lt")
Similarly, the "Data Dihedrals" and "Data Improper" sections contain 6 columns, and include the @dihedral and @improper types.
```
write("Data Dihedrals") {
   $dihedral:d1 @dihedral:TYPE1 $atom:H30 $atom:C18 $atom:C17 $atom:C16
   $dihedral:d2 @dihedral:TYPE2 $atom:H30 $atom:C18 $atom:C17 $atom:C19
    :
}
```
... But again, there is no reason to include this information, because the OPLSAA force field will create these angles, dihedrals, and imrpopers for you.  That's why I suggest deleting the Angles, Dihedrals, and Impropers from your "D7JH_allatom_optimized_geometry.lt" file.

I apologize that "moltemplate.sh" is such a complicated and confusing program.

At least I will modify the error message so that future moltemplate users will not suffer as much as you did.

Hope this helps.

Andrew
I'll close the issue soon once I fix the error message.

**Comment by jewettaij** [comment]
(I'm glad more people are using ATB.  I've been getting more questions about them lately.)

**Comment by dihedral** [comment]
appreciate the ping ;-) :-P

**Comment by davidfir3** [comment]
@jewettaij Thanks for your detailed explanation. And Sorry for the mistakes in my files. I took it for granted that there were similar "Data Angle List" sections coupled with "Data Angles By Type", as "Data Bond List" sections do. I'll delete them and try again.

As for your question, I observed an undesirable behavior of molecules in my simulation, guessing that it might result from the bad choice of force field. So I wanted to try OPLSAA, and then met with this issue. (Later, I realized that force field was not the problem.)

I really appreciate your efforts of developing such a convenient tool. It helps a lot! And I wish it will be better. Thank you again.

**Comment by davidfir3** [comment]
@jewettaij Thanks for your detailed explanation. And Sorry for the mistakes in my files. I took it for granted that there were similar "Data Angle List" sections coupled with "Data Angles By Type", as "Data Bond List" sections do. I'll delete them and try again.

As for your question, I observed an undesirable behavior of molecules in my simulation, guessing that it might result from the bad choice of force field. So I wanted to try OPLSAA, and then met with this issue. (Later, I realized that force field was not the problem.)

I really appreciate your efforts of developing such a convenient tool. It helps a lot! And I wish it will be better. Thank you again.
